### ISSUE TITLE
linter: Rework only_used_in_recursion detection logic

### ISSUE THREAD

**Issue by no-yan** [oxc-project/oxc#7275]
### What happened?

To improve detection, I propose modifying the algorithm to support a wider range of patterns, including multi-node expressions and cases with more complex references.

The current detection logic is conservative and based on the following conditions:
1. All references are used in the appropriate position (property or argument) of the recursive function call.
2. The call is a single-node invocation, e.g., `recurfunc(arg)`, and not a multi-node expression, e.g., `recurfunc(arg + 1)`.

Due to this, the following cases are not detected by the current implementation:
```javascript
function recursive(arg) {
    return recursive(arg + 1);
}
```

```javascript
function recursive(arg) {
    console.log(arg);
    return recursive(arg);
}
```

[Clippy's implementation](https://github.com/Jarcho/rust-clippy/blob/39f4bee98e2ad69bb247b57b6d5bc8df0b4f1090/clippy_lints/src/only_used_in_recursion.rs#L294-L350) avoids issuing warnings for expressions with side effects.  
We might consider adopting a similar approach in our implementation.

**Comment by Boshen** [comment]
@no-yan would like to work on this, or should I change this to good first issue?

**Comment by no-yan** [comment]
@Boshen Unfortunately, I didn't have the bandwidth to work on this for now.
Could you please unassign me?

**Comment by Boshen** [comment]
Close as stale.
